app/src/main/java/net/programmierecke/radiodroid2/alarm/AlarmReceiver.java

```diff
@@ -199,4 +203,72 @@ protected void onPostExecute(String result) {
             }
         }.execute();
     }
+
+    int currentVolume;
+    int minVolume;
+    private void graduallyIncreaseAlarmVolume(final Context context, boolean checkIfPlaying) {
+        SharedPreferences sharedPref = PreferenceManager.getDefaultSharedPreferences(context);
+        int slowWakeMillis = Integer.parseInt(sharedPref.getString("gradually_increase_volume", "0"));
+
+        AudioManager audioManager = (AudioManager) context.getSystemService(Context.AUDIO_SERVICE);
+        int originalVolume = audioManager.getStreamVolume(AudioManager.STREAM_ALARM);
+
+        if (slowWakeMillis == 0 || originalVolume == 0) {
+            if (BuildConfig.DEBUG) { Log.d(TAG, "Gradual alarm volume disabled"); }
+            return;
+        }
+
+        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.P) {
+            minVolume = audioManager.getStreamMinVolume(AudioManager.STREAM_ALARM);
+        } else {
+            minVolume = 0;
+        }
+
+        int volumeRange = originalVolume - minVolume;
+        currentVolume = minVolume;
+
+        audioManager.setStreamVolume(AudioManager.STREAM_ALARM, minVolume, 0);
```

Comment:
Changing global volume for the stream is a wrong way. There is `RadioPlayer.setVolume`.

If you want to cover the case with external player, better leave out for now. I'm quite reserved about messing with global volume.

Reply:
👍 Good idea, while I was testing it was sometimes leaving the system alarm volume at the wrong level, I've updated this to change the volume via the `PlayerService` in fb4973930c15c5ead97783cce466092af838bfa1...05330079cb11e1f6d459ecaafe55c3a3cc1a0767.